hudi-client/hudi-client-common/src/main/java/org/apache/hudi/metadata/index/IndexerFactory.java

```diff
@@ -0,0 +1,107 @@
+import org.apache.hudi.util.Lazy;
+
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.Map;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+import static org.apache.hudi.metadata.MetadataPartitionType.getValidValues;
+
+public class IndexerFactory {
+  public static Indexer getIndexer(MetadataPartitionType partitionType,
+                                   HoodieEngineContext engineContext,
+                                   HoodieWriteConfig dataTableWriteConfig,
+                                   HoodieTableMetaClient dataTableMetaClient,
+                                   Lazy<HoodieTable> table,
+                                   ExpressionIndexRecordGenerator expressionIndexRecordGenerator) {
+    switch (partitionType) {
+      case BLOOM_FILTERS:
+        return new BloomFiltersIndexer(
+            engineContext, dataTableWriteConfig, dataTableMetaClient);
+      case COLUMN_STATS:
+        return new ColumnStatsIndexer(
+            engineContext, dataTableWriteConfig, dataTableMetaClient);
+      case EXPRESSION_INDEX:
+        return new ExpressionIndexer(
+            engineContext, dataTableWriteConfig, dataTableMetaClient, expressionIndexRecordGenerator);
+      case FILES:
+        return new FilesIndexer(engineContext);
+      case PARTITION_STATS:
+        return new PartitionStatsIndexer(
+            engineContext, dataTableWriteConfig, dataTableMetaClient);
+      case RECORD_INDEX:
+        return new RecordIndexer(
+            engineContext, dataTableWriteConfig, dataTableMetaClient, table);
+      case SECONDARY_INDEX:
+        return new SecondaryIndexer(
+            engineContext, expressionIndexRecordGenerator.getEngineType(), dataTableWriteConfig, dataTableMetaClient);
+      default:
+        throw new HoodieNotSupportedException(
+            "Unsupported metadata partition type for indexing: " + partitionType);
+    }
+  }
+
+  /**
+   * @param engineContext                  {@link HoodieEngineContext} instance
+   * @param dataTableWriteConfig           write config for the data table
+   * @param dataTableMetaClient            meta client of the data table
+   * @param table                          lazy {@link HoodieTable} instance presenting the data table
+   * @param expressionIndexRecordGenerator record generator for the expression index
+   * @return the map of metadata partition type to the indexer for the enabled metadata
+   * partition types based on the metadata config and table config.
+   */
+  // TODO(yihua): remove MetadataPartitionType#getEnabledIndexBuilderMap
+  public static Map<MetadataPartitionType, Indexer> getEnabledIndexerMap(
+      HoodieEngineContext engineContext,
+      HoodieWriteConfig dataTableWriteConfig,
+      HoodieTableMetaClient dataTableMetaClient,
+      Lazy<HoodieTable> table,
+      ExpressionIndexRecordGenerator expressionIndexRecordGenerator) {
+    if (!dataTableWriteConfig.getMetadataConfig().isEnabled()) {
```

Comment:
guess you gonna fix this for secondary index and EI right from what we discussed.